Author review (#63)
gammapy/gammapy-v1.0-paper: src/ms-review.tex

\gammapy provides its user community with a tested and versioned up-to-date
on-line
documentation\footnote{\url{https://docs.gammapy.org}}~\citep{2019ASPC..523..357B}
built with Sphinx\footnote{\url{https://www.sphinx-doc.org}} scanning the
codebase Python scripts, as well as a set of RST files and Jupyter notebooks.
The documentation includes a handwritten user guide, a set of executable
tutorials, and a reference to the API automatically extracted from the code and
docstrings. The \gammapy code snippets present in the documentation are tested
in different environments using our continuous integration (CI) workflow based
on \github actions.

The Jupyter notebooks tutorials are stored stripped of their output cells,
which greatly helps in identifying differences in the contribution review.
During the CI these are executed, the output cells parsed and, in case an error
is found in one of the output cells then validation fails. Code formatting of
the input cells is also done, parsing their content and using a particular
syntax to provide links to the API documentation pages. The resulting web
published tutorials also provide links to playground spaces in

> myBinder~\citep{project_jupyter-proc-scipy-2018}, where they may be executed

Author: Binder will be working for v1.0...so we just keep it in the paper?

API design. This will allow users to define arbitrary complex analysis
scenarios as YAML files and even extend their workflows by user defined
analysis steps via a registry system. Another important topic will
be to improve the support of handling meta data for data structures
and provenance information to track the history of the data reduction
process from the DL3 to the highest DL5/DL6 data levels.

Around the core Python package a large diverse community of
users and contributors has developed. With regular developer meetings,
coding sprints and in-person user tutorials at relevant conferences
and collaboration meetings, the community has constantly grown.
So far \gammapy has seen ~80 contributors from ~10 different countries.
With typically ~10 regular contributors at any given time of the
project, the code base has constantly grown its range of features
and improved its code quality. With \gammapy being selected as the base library
for the future science tools for CTA, we expect the community to grow
even further, providing a stable perspective for further usage,
development and maintenance of the project. Besides the future use
by the CTA community \gammapy has already

> been used for analysis of data from the \hess and \magic instruments

Author: Done

While \gammapy was mainly developed for the science community around

> IACT instruments, the data model and software design is general

Reviewer: ```suggestion
IACT instruments, the internal data model and software design are general
```

enough to be applied to other \gammaray instruments as well.
The use of \gammapy for the analysis of data from the High Altitude
Water Cherenkov Observatory (HAWC) has been successfully
demonstrated by \todo{Olivera et al.}. This makes \gammapy
a viable choice for the base library for the science tools
of the future Southern Widefield Gamma Ray Observatory

> (SWGO) and use with data from LHAASO as well. \gammapy

Reviewer: first references of these 2 experiments. References are needed...

> (SWGO) and use with data from LHAASO as well. \gammapy

Reviewer: ```suggestion
(SWGO) and use with data from Large High Altitude Air Shower Observatory (LHAASO) as well. \gammapy
```

> (SWGO) and use with data from LHAASO as well. \gammapy

Reviewer: Proposal of addition:
As very high-energy neutrino observatories share the same analysis flow as the non-pointing instruments like HAWC, SWGO and LHAASO, the Cubic Kilometre Neutrino Telescope (KM3NeT) community is also experiencing \gammapy, which would extend it to a multi-messenger analysis tool for the very high-energy astrophysics.

Author: Done

	Universidad Complutense de Madrid EMFTEL, UCM, Plaza de la Ciencias, 28040 Madrid, Spain \label{inst:26} \and
	Universidade de São Paulo, Brazil \label{inst:27} \and
	Université Savoie Mont Blanc, CNRS, Laboratoire d’Annecyde Physique des Particules - IN2P3, 74000 Annecy, France \label{inst:28} \and
	unknown \label{inst:unknown}
}

% \abstract{}{}{}{}{}
% 5 {} token are mandatory

\abstract{
	% TODO: the feature are still to detailed and partly copy and pasted form the ICRS abstract...avoid this
	% context heading (optional)
	% {} leave it empty if necessary
	%{Gammapy context:}
	Historically the data as well as analysis software in \gammaray astronomy is
	proprietary to the experiments. With the future Cherenkov Telescope Array
	(CTA), which will be operated as an open \gammaray observatory with public
	data, there is a corresponding need for open high-level analysis software. In
	this article we present the first major version v1.0 of \gammapy, a
	community-developed open-source Python package for \gammaray astronomy. We
	present its general design and provide an%
	overview of the analysis methods and features it implements. Starting from

> 	event lists and a description of the specific instrument response functions

Author: I think "specific" refers to instrument here, maybe just switch like "instrument specific response functions"?

> 	(IRF) stored in open FITS based data formats, \gammapy implements . Thereby it

Author: Fixed

	handles the dependency of the IRFs with time, energy as well as position on the
	sky. It offers a variety of background estimation methods for spectral, spatial
	and spectro-morphological analysis. Counts, background and IRFs data are
	bundled in datasets and can be serialised, rebinned and stacked. \gammapy
	supports to model binned data using Poisson%
	maximum likelihood fitting. It comes with built-in spectral, spatial and
	temporal models as well as support for custom user models, to model e.g.,
	energy dependent morphology of \gammaray sources. Multiple datasets can be
	combined in a joint-likelihood approach to either handle time dependent IRFs,
	different classes of events or combination of data from multiple instruments.

> 	Gammapy also implements methods to estimate flux points, including likelihood

Author: Done

	% The \gammaray can be considered as the last frontier
	% In the last two decades the measurement of \gammaray emission from the universe has
	% evolved from a niche

	% aims heading (mandatory)
	%{Gammapy aims}
	%Reproducible analyses, share algorithms etc.

	% methods heading (mandatory)
	%{Gammapy methods}

	% results heading (mandatory)
	%{Gammapy results}

	% conclusions heading (optional), leave it empty if necessary
	%{Gammapy conclusions}
}

% \date{Received September 15, 1996; accepted March 16, 1997}
\keywords{
	Gamma rays: general -
	Astronomical instrumentation, methods and techniques -
	Methods: data analysis
}

\maketitle

% Main part
\section{Introduction}
\label{sec:introduction}

%% \gammaray astronomy
\gammaray astronomy is a rather young field of research. The \gammaray~
range of the electromagnetic spectrum provides us with insight to the
most energetic process in the universe such as surroundings of
black holes and remnants of supernova explosions. As in other
branches of astronomy, \gammarays can be observed by both
satellite as well as ground based instruments.
Ground based instrument use the Earth's atmosphere as a particle detector.

> Very high energy cosmic \gammarays interact in the atmosphere and

Reviewer: ```suggestion
Very high-energy cosmic \gammarays interact in the atmosphere and
```

good energy and angular resolution.

\begin{figure*}[t]
	\centering
	\includegraphics[width=\textwidth]{figures/big-picture.pdf}
    \caption{
        Relation of \gammapy to different \gammaray instruments and the data formats.
        \gammapy provide access to the data stored in GADF format and the
        high-level data analysis functionality.}
	\label{fig:big_picture}

\end{figure*}

Water Cherenkov observatories detect directly particles from the tail of the
shower when it reaches the ground. These instruments have very
large field-of-view, large duty-cycle but higher energy threshold and
usually have lower signal to noise ratios compared to IACTs.

%% Context
Ground based \gammaray astronomy has been historically structured
by experiments run by independent collaborations relying
on their own proprietary data and analysis software developed as part of the
instrument. While this model has been very successful so far, it does not
permit easy combination of data from several instruments and is therefore
a limitation for interoperability of existing facilities. Especially
because the different detection techniques have complementary

> properties.

Author: Addressed

%TODO: better transition into context
The Cherenkov Telescope Array (CTA) will be the first instrument to be operated
as an open observatory in the domain. Its high level data will be shared publicly after
some proprietary period and the software required to analyze it will be distributed
as well. To allow the re-usability of existing instruments and their interoperability
requires open data formats and open tools that can support the various analysis methods
commonly used in the field.

%% Context : formats
In practice, the data-reduction workflow of all \gammaray observatories
is remarkably similar. After data calibration, shower events are reconstructed and
gamma/hadron separation is applied to build lists of \gammaray like events.
The latter are then used to derive scientific results, such as spectra, sky maps

> or light curves, taking into account the specific instrument response functions (IRF).

Author: Done

The information in this high data level is independent on

> the data reduction, and eventually of the detection technique. This implies,

Author: Clarified

for example, that data from IACT and particle samplers can be represented
within the same model. The efforts to prototype a format usable by various instruments
converged in the so-called \textit{Data Format for \gammaray Astronomy}
initiative~\citep{gadf_proc, gadf_universe}, abbreviated in
\texttt{gamma-astro-data-format} (\gadf). The latter proposes prototypical
specifications to produce files based on the flexible image transport system
(FITS) format~\citep{fits} encapsulating this high-level information. This is
realized by storing a list of gamma-like events with their measured quantities
(energy, direction, arrival time) and a parametrisation of the response of the
system.

%% Context: develoment of open software
Python has become extremely popular as a scientific programming language
in particular in the field of data sciences. The success is
mostly attributed to the simple and easy to learn syntax, the ability to act as

> a "glue" language between different programming languages and last but not least

Reviewer: ```suggestion
a 'glue' language between different programming languages and last but not least
```

the rich eco-system of packages and its open and supportive community, from
core numerical analysis packages such as \numpy~\citep{numpy} and
\scipy~\citep{2020SciPy-NMeth}, visualization libraries,e.g. Matplotlib \citep{matplotlib},
or optimization packages such as Iminuit \citep{iminuit}.

The \astropy~\citep{astropy} was created in 2012 to build a community-developed
Python package for astronomical reasearch. Among other things, the it offers functionality for
transforming and representing astronomical coordinates, manipulating physical quantities
and their units as well as reading and writing FITS files.

%% \gammapy: concept and goals
%%% Question: starting date? Is is relevant, what is the real start?
The \gammapy project started in 2015 with the objective of building a flexible and

> efficient software library for very high energy \gammaray data analysis \citep{gammapy_2015}.

Reviewer: ```suggestion
efficient software library for very high-energy \gammaray data analysis \citep{gammapy_2015}.
```

This is possible thanks to the definition of a common data format such as \gadf, and thanks to
an open development with a broad community of contributors. The relation of \gammapy to the

> different instruments the \gadf data format is illustarted in Figure~\ref{fig:big_picture}.

Reviewer 2: ```suggestion
different instruments and the \gadf data format is illustarted in Figure~\ref{fig:big_picture}.
```

%\gammapy is built on the scientific Python stack and makes use of Astropy.

%% \gammapy: from first developments to validation and selection
The \hess collaboration released a limited test dataset (about 50 hours of
observations taken between 2004 and 2008) based  on the \gadf DL3 format \citep{HESS_DR1}.
This data release served as a basis for analysis tools validation \cite[see e.g.]{Mohrmann2019}.

%% Paper outline
In this article, we describe the general structure of the \gammapy package and
its main concepts. In Section \ref{sec:analysis-workflow-overview}, we present

> the data analysis workflow in very high energy \gammaray astronomy. Then we

Reviewer: ```suggestion
the data analysis workflow in very high-energy \gammaray astronomy. Then we
```

based of their meta data including information such as pointing direction, observation
time and observation conditions.

The next step of the analysis is the data reduction where all observation events and instrument
responses are projected onto a user-defined geometry.A typical geometry consists in a spectral representation with a measured
energy axis, and in a spatial representation, either a coordinates system  with a projection
(for 3-dimensional or cube analysis) or a region on the sky (for regular spectral analysis).
The gammapy.maps subpackage provides general multidimensional geometry objects
(Geom) and the associated data structures (Maps), see \ref{ssec:gammapy-maps}.

All observation events and instrument responses are projected onto the
user defined geometry. Because residual hadronic background models can be subject
to significant uncertainties, background correction must be applied,
such as the ring or the field-of-view background techniques or
background measurements must be performed within, e.g. reflected regions~\citep{Berge07}.
Parts of the data with high associated IRF systematics must also be excluded by defining

> a "safe" data range. These data reduction steps are performed by classes and functions

Reviewer: ```suggestion
a 'safe' data range. These data reduction steps are performed by classes and functions
```

tables, with quantities expressed in terms of energy (true or reconstructed),
off-axis angles or cartesian detector coordinates. The main information stored in
the common \gammaray \irfs are the effective area (Aeff), energy dispersion
(Edisp), point spread function (PSF) and background rate (BKG). The \code{gammapy.irf}
sub-package can open and access specific \irf extensions,
interpolate and evaluate the quantities of interest on both energy and spatial
axes, convert their format or units in different kinds, plot or write them into
output files. In the following, we list the main classes of the
sub-package:

\paragraph{Effective area:}
\gammapy provides the class \code{EffectiveAreaTable2D} to
manage Aeff, which is usually defined in terms of true energy and offset angle.
The class functionalities offer the possibility to read from files or to create
it from scratch. {\code{EffectiveAreaTable2D} can also convert, interpolate,
write, and evaluate the effective area for a given energy and offset angles, or
even plot the multi-dimensional Aeff table.

\paragraph{Point spread function:}
\gammapy allows user to treat different kinds of PSFs,

> in particular, parametric multi-dimensional gaussians (\code{EnergyDependentMultiGaussPSF})

Reviewer: ```suggestion
in particular, parametric multi-dimensional Gaussians (\code{EnergyDependentMultiGaussPSF})
```

or King profile functions (\code{PSFKing})
one. The \code{EnergyDependentMultiGaussPSF} class is able to handle up to three

> gaussians, defined in terms of amplitudes and sigma given for each true energy

Reviewer: ```suggestion
Gaussians, defined in terms of amplitudes and sigma given for each true energy
```

Author: Done

and offset angle bin. Similarly, \code{PSFKing takes} into account the gamma and
sigma parameters as defined here. The \code{ParametricPSF} allows to create a

> PSF with a representation different from gaussian(s) or King profile(s).

Reviewer: ```suggestion
PSF with a representation different from Gaussian(s) or King profile(s).
```

Finally, the user can take advantage from the \code{PSFMap} class, which creates
a multi-dimensional map of the PSF in WCS coordinates. At each position, a PSF
kernel map (\code{PSFKernel}) provides a PSF as a function of the true energy.
The creation of PSF kernel maps, where the PSF is defined for each
sky-position, is also given to the user. The latter two can speed up analyses.

> \paragraph{Energy dispersion:}

Reviewer: There is no notion of probability density as mentioned in section 2. To be precised..
And, Edisp should be stored as probability or we renormalise it internally?

Edisp, in \iact, is generally given in terms of the so
called migration parameter ($\mu$), which is defined as the ratio between the
reconstructed energy and the true energy. This ratio should be as close as one

> and its dispersion can assume the shape of a gaussian (or even more complex

Reviewer: Gammapy assume? why 'or more complex shape'? do we use a random shape or not? This sentence needs clarification

> and its dispersion can assume the shape of a gaussian (or even more complex

Reviewer: ```suggestion
and its dispersion can assume the shape of a Gaussian (or even more complex
```

> distributions). Migration parameter is given at each offset angle and

Reviewer: See my previous comment on 'offset'

reconstructed energy. The main sub-classes are \code{EnergyDispersion2D},
designed to interpret Edisp, \code{EDispKernelMap}, which builds an Edisp kernel
map, i.e., a 4-dimensional WCS map where at each sky-position is associated an
Edisp kernel. The latter is a representation of the Edisp as a function of the
true energy only thanks to the sub-class \code{EDispKernel}.

\paragraph{Background model:}

> The BKG can be represented in \gammapy as either 1) a 2D map (\code{Background2D})

Reviewer: The 1) and 2) are not necessary needed. But this is just a matter of style;)

of count rate normalised per steradians and energy at different

> energies and offset-angles or 2) as rate per steradians and energy, as a

Reviewer: ```suggestion
reconstructed energies and offset angles or 2) as rate per steradians and energy, as a
```

function of reconstructed energy and detector coordinates (\code{Background3D}).
In the former, the background is expected to follow a radially symmetric shape,
while in the latter, it can be more complex.

\subsection{gammapy.maps}
\label{ssec:gammapy-maps}
The \code{gammapy.maps} sub-package provides classes that represent data
structures associated with a set of coordinates or a region on a sphere. In
addition it allows to handle an arbitrary number of non-spatial data
dimensions, such as time or energy. It is organized around three types of

> structures: geometries, sky-maps and map axes, which inherit from the base

Reviewer: ```suggestion
structures: geometries, sky maps and map axes, which inherit from the base
```

classes \code{Geom}, \code{Map} and \code{MapAxis} respectively.

The geometry object defines the pixelization scheme and map boundaries. It also
provides methods to transform between sky and pixel coordinates. Maps consist

> of a geometry instance together with a Numpy data array containing the corresponding

Reviewer: First mention of Numpy! need a reference here...

The publication scheme of Numpy is here: [citing Numpy](https://numpy.org/citing-numpy/)

map values. Map axes contain a sequence of ordered values which define bins on
a given dimension, spatial or not. Map axes can have physical units attached to
them, as well as define non-linear spaced bins. All map classes support operations such
as arithmetic operations with unit support, up- and downsampling along
extra axes, interpolation, resampling of extra axes, interactive visualisation
in notebooks and interpolation onto different geometries.

> The sub-package provides a uniform API for the FITS World Coordinate System

Reviewer: API is a term that never have been defined up to here. To be done...

(WCS), the HEALPix pixelization and region-based data structure
(see Figure~\ref{ig*:minted:gp_maps}).

\begin{figure}
	\import{code-examples/generated/}{gp_maps}

	\caption{
        Using \code{gammapy.maps} to create a WCS, a HEALPix and a region
		based data structures. The initialisation parameters include
        consistently the positions of the center of the map, the pixel
        size, the extend of the map as well as the energy axis definition.
        The energy minimum and maximum values for the creation of the
        \code{MapAxis} object can be defined as strings also specifying the
        unit. Region definitions can be passed as strings following
        the DS9 region specifications \url{http://ds9.si.edu/doc/ref/region.html}.
        }
    \label{ig*:minted:gp_maps}
\end{figure}

> % itemize because it helps me write, could also just be paragraphs...

Reviewer: the `subsubsection` creates a `\newline` between the 3. And there is not just before the paragraph of the `MapAxis`.
Maybe a `\paragraph{XXX:}` is better, and taking care that we have a `\newline` before the `MapAxis` paragraph...

\subsubsection{WCS Maps}
The FITS WCS pixelization supports a different number of projections to
represent celestial spherical coordinates in a regular rectangular grid.
\gammapy provides full support to data structures using this pixelization
scheme. For details see ~\cite{Calabretta2002}. This pixelisation
is typically used for smaller regions of interests, such as pointed
observations.

\subsubsection{HEALPix Maps}
This pixelization scheme ~\citep{Calabretta2002} provides a
subdivision of a sphere in which each pixel covers the same surface area as
every other pixel. As a consequence, however, pixel shapes are no longer
rectangular, or regular.
This pixelisation is typically used for all-sky data, such as data
from the \hawc or \fermi observatory. \gammapy natively supports
the multiscale definiton of the HEALPix pixelisation and thus
allows for easy up and downsampling of the data. In addition to
the all-sky map, \gammapy also supports a local HEALPix
pixelisation where the size of the map is constrained to a given
radius.

> For local nighbourhood operations, such as convolution \gammapy relies

Reviewer: ```suggestion
For local neighbourhood operations, such as convolution \gammapy relies
```

on projecting the HEALPix data to a local tangential WCS grid.

\subsubsection{Region Maps}
In this case, instead of a fine spatial grid
dividing a rectangular sky region, the spatial dimension is reduced to a single
bin with an arbitrary shape, describing a region in the sky with that same
shape. Typically they are is used together with a non-spatial dimension, for
example an energy axis, to represent how a quantity varies in that dimension
inside the corresponding region. The region is represented on the local
tangential projection see \cite{Astropy regions}.

Additionally, the \code{MapAxis} class provides a uniform API for axes representing
bins on any physical quantity, such as energy or angular offset. The special
case of time is covered by the dedicated \code{TimeMapAxis}, which allows time bins to
be non-contiguous, as it is often the case with observation time-stamps. The
generic class \code{LabelMapAxis} allows the creation of axes for non-numeric
entries.

> \subsection{gammapy.datasets}

Reviewer: For a non-expert, this paragraph is hard to read... and for Atreyee hard to write, because one speaks about statistics, Cstat, Wstat, likelihood, etc without defining it. In the reading flow, it is not great...
I am wondering whether one should/could move the sub-section `gammapy.stats` just before this sub-section? Then it could be simplied a bit

\label{ssec:gammapy-datasets}
%
\begin{figure}
	\import{code-examples/generated/}{gp_datasets}
	\caption{
        Using \code{gammapy.datasets} to read existing reduced binned datasets.
        After the different datasets are read from disk they are collected into a
        common \code{Datasets} container. All dataset types have an associated
        name attribute to allow access by name later in the code. The
        environment variable \code{\$GAMMAPY\_DATA} is automtically resolved
        by \gammapy.
    }
	\label{fig*:minted:gp_datasets}
\end{figure}
%
The end product of the data reduction process described in Section

> \ref{ssec:gammapy-makers} are a set of binned counts, background and IRF maps,

Reviewer: ```suggestion
\ref{sec:gammaray-data-analysis} (and in detail in \ref{ssec:gammapy-makers}) are a set of binned counts, background and IRF maps,
```

> \ref{ssec:gammapy-makers} are a set of binned counts, background and IRF maps,

Reviewer: the backgound is an IRF (see the corresponding sub-section).... May I propose "binned counts and IRF maps (such as the background map),

at the DL4 level. The \code{gammapy.datasets} subpackage contains classes to bundle

> together binned data along with associated models and the likelihood, which

Reviewer: you does not mean the function, but its value and the name of the used statistics?

> provides an interface to the Fit class (Sec \ref{ssec:gammapy-modeling}) for

Reviewer: ```suggestion
provides an interface to the \code{Fit} class (Sec \ref{sssec:fit}) for
```

modeling and fitting purposes. Depending upon the type of analysis and the
associated statistics, different types of Datasets are supported. \code{MapDataset} is
used for 3D (spectral and morphological) fitting, and a 1D spectral fitting is
done using \code{SpectrumDataset}. While the default statistics for both of these is
Cash, their corresponding on off versions are adapted for the case where the

> background is measured from real off counts, and support wstat statistics. The

Reviewer: See my general comment about this sub-section

predicted counts are computed by convolution of the models with the associated
IRFs. Fitting of precomputed flux points is enabled through \code{FluxPointsDataset},
using chi2 statistics. Multiple datasets of same or different types can be
bundled together in \code{Datasets} (e.g., Figure \ref{fig*:minted:gp_datasets}),
where the likelihood from each constituent member is added, thus facilitating
joint fitting across different observations, and even different instruments
across different wavelengths. Datasets also provide functionalities for

> manipulating reduced data, eg: stacking, sub-grouping, plotting, etc. Users can

Reviewer: ```suggestion
manipulating reduced data, e.g. stacking, sub-grouping, plotting. Users can
```

also create their customized datasets for implementing modified likelihood
methods.

\subsection{gammapy.makers}
\label{ssec:gammapy-makers}
%
\begin{figure}
    \import{code-examples/generated/}{gp_makers}
	\caption{
        Using \code{gammapy.makers} to reduce DL3 level data into a
		\code{MapDataset}. All \code{Maker} classes take
        the configuration on initialisation of the class.
    }
	\label{ig*:minted:gp_makers}
\end{figure}
%
The data reduction step includes all tasks required to process and prepare
\gammaray data from the DL3 level to modeling and fitting. The \code{gammapy.makers} sub-package
contains the various classes and functions required to do so. First, events are
binned and IRFs are interpolated and projected onto the chosen analysis
geometry. This produces counts, exposure, background, psf and energy dispersion
maps. The \code{MapDatasetMaker} and \code{SpectrumDatasetMaker} are
responsible for this task, see Fig~\ref{ig*:minted:gp_makers}.

Because the background models suffer from strong uncertainties it is required
to correct them from the data themselves. Several techniques are commonly used

> in \gammaray astronomy such as field-of-view background normalization or

Reviewer: ```suggestion
in TeV \gammaray astronomy such as field-of-view background normalization or
```

background measurement in reflected regions, see~\cite{Berge07}.
Specific \code{Makers} such as the \code{FoVBackgroundMaker} or the
\code{ReflectedRegionsBackgroundMaker} are in charge of this step.

Finally, to limit other sources of systematic uncertainties, a data validity
domain is determined by the \code{SafeMaskMaker}. It can be used to limit the

> extent of the field of view used or to limit the energy range to e.g., a domain

Reviewer: ```suggestion
extent of the field of view used or to limit the energy range to, e.g., a domain
```

models, \gammapy uses a factorised representation of the total source
model:

\begin{equation}
    f(l, b, E, t) = F(E)  \cdot G(l, b, E) \cdot  H(t, E)
\end{equation}

Where the spectral model $F$ implicitly carries the total flux
norm of the model. The spatial model $G$ optionally depends
on energy and allows to support e.g. energy dependent
morphology models. The temporal model component $H$
optionally also supports an energy variable to
allow for spectral variations of the model with time.

Following the factorisation The models objects are grouped
into the following categories:

\begin{itemize}
	\item \code{SpectralModel}: models to describe spectral shapes of sources
	\item \code{SpatialModel}: models to describe spatial shapes (morphologies) of sources
	\item \code{TemporalModel}: models to describe temporal flux evolution of sources, such as
	      light and phase curves
\end{itemize}

The models follow a naming scheme which contains the category as a suffix to
the class name.

The spectral models include a special class of normed models, which have a
dimension-less normalisation. These spectral models feature a norm parameter

> instead of amplitude and are named using the NormSpectralModel suffix. They

Author: Fixed.

must be used along with another spectral model, as a multiplicative correction
factor according to their spectral shape. They can be typically used for

> adjusting template based models, or adding a EBL correction to some analytic

Author: Addressed

built-in. We also offer a wrapper to radiative models implemented in the Naima
package~\cite{naima}. The modeling framework can be easily extended with
user-defined models. For example agnpy models that describe leptonic radiative
processes in jetted Active Galactic Nuclei (AGN) can wrapped into
\gammapy~\citep[see section3.5 of ][]{2021arXiv211214573N} .

\begin{figure}
	\import{code-examples/generated/}{gp_models}
	\caption{Using \code{gammapy.modeling.models} to define a source model with a
    spectral, spatial and temporal component. For convenience the model
    parameters can be defined as strings with attached units. The spatial model
    takes an addtional \code{frame} parameter which allow users to define
    the coordinate frame of the position of the model.
    }
	\label{fig*:minted:gp_models}
\end{figure}

\subsubsection{Fit}
\label{sssec:fit}

The \code{Fit} class provides methods to fit i.e., optimise parameters and estimate
parameter errors and correlations. It interfaces with a Datasets object, which
in turn is connected to a Models object containing the model parameters in its
Parameters object. Models can be unique for a given dataset, or contribute to

> multiple datasets and thus provide links, allowing e.g., to do a joint fit to

Reviewer 2: ```suggestion
multiple datasets, allowing e.g., to do a joint fit to
```

multiple IACT datasets, or to a joint IACT and \fermi dataset. Many
examples are given in the tutorials.

\todo{Add fit code example...?}

The Fit class provides a uniform interface to multiple fitting backends:
“minuit”~\citep{iminuit}, “scipy”,~\citep{2020SciPy-NMeth}, and
“sherpa”~\citep{sherpa-2005,sherpa-2011}. Note that, for now, covariance matrix
and errors are computed only for the fitting with MINUIT. However depending on
the problem other optimizers can better perform, so sometimes it can be useful
to run a pre-fit with alternative optimization methods. In future we plan to
extend the supported fitting backends, including for example MCMC solutions.
\footnote{a prototype is available in gammapy-recipes,
	\url{https://gammapy.github.io/gammapy-recipes/_build/html/notebooks/mcmc-sampling-emcee/mcmc_sampling.html}
}

\subsection{gammapy.estimators}
\label{ssec:gammapy-estimators}
By fitting parametric models to the data, the total integrated

> flux and overall temporal, spectral and morphological shape of the

Author: I would argue here a single source can only have a single shape (in an abstract sense...)

\gammaray emission can be constrained. In many cases it useful
to make a more detailed follow-up analysis by measuring the
flux in smaller spectral, temporal or spatial bins. This
possibly reveals more detailed emission features, which
are relevant for studying correlation with counterpart emissions.

The \code{gammapy.estimators} sub-module features methods to compute flux
points, light curves, flux maps, flux profiles from data.
The basic method for all these measurements is equivalent.
The initial fine bins of \code{MapDataset} are grouped into
larger bins. The norm of the best fit "reference" spectral

> model is fitted ("scaled") in the  restricted data range only.

Reviewer: ```suggestion
model is fitted (or 'scaled') in the  restricted data range only.
```

The base class of all algorithms is the \code{Estimator}  class.

Signifiance is estimated based on hypothesis
testing against the background only model. This allows
to assign a significance to the given flux bin.
In addition to the best fit flux norm all estimators compute
quantities corresponding to this flux. This includes
the predicted number of total, signal and background
counts per flux bin. The total fit statistics
of the best fit model, the fit statistics of the
null hypothesis and the difference between both,
the so-called TS value.
From the TS value the significance of the
significance of the measured signal and associated flux
can be derived.

Optionally it can also compute more advanced quantities
such as assymmetric flux errors, flux upper limits
and one dimensional profiles of the fit statistic,

> which show how the likelihood functions varies with

Reviewer: ```suggestion
which shows how the likelihood function varies with
```

the flux norm parameter around the fit minimum.
This information is useful for inspecting the quality

> of the fit, where assumptoticcally a parabolic

Reviewer: ```suggestion
of a fit, for which asymptomatically a parabolic
```

shape of the profile is expected at the best fit

> values (\todo{Reference to modeling / fitting section?}).

Reviewer: I have even not sure that in our documentation we speak about this statistical properties... Maybe one could create an issue for the document.
The demonstration can be found in:
- Wilks 1962 ($13.8)
- Eadie et al. 1971 ($10.5.2)
- Li 1980 ($6.3.4)
I found these references in the Li&Ma paper.

The result of the flux point estimation are either stored in a
\code{FluxMaps} or \code{FluxPoints} object. Both objects
are based on an internal representation of the flux which is

> independent of the SED type. The flux is represented by a

Reviewer: ```suggestion
independent of the Spectral Energy Distribution (SED) type. The flux is represented by a
```

the reference spectral model and an array of
normalisation values given in energy, time and spatial bins,

> which factorise the deviation of the flux in the given

Reviewer: ```suggestion
which factorises the deviation of the flux in a given
```

> bin form the reference spectral model. This allows

Reviewer: ```suggestion
bin for the reference spectral model. This allows
```

user to conveniently transform between different
SED types. Table~\ref{tab:sed_types} shows an
overview and definitions of the supported SED types.
The actual flux values for each SED type are obtained
by mutiplication of the norm with the  reference flux.

\begin{table*}
    \begin{center}
        \begin{tabular}{lll}
         \hline
         Type & Description & Unit Equivalency \\
         \hline

>          dnde & Differential flux at a given energy & $\mathrm{TeV^{-1}~cm^{2}~s^{-1}}$ \\

Reviewer: ```suggestion
         dnde & Differential flux at a given energy & $\mathrm{TeV^{-1}~cm^{-2}~s^{-1}}$ \\
```

>          e2dnde & Differential flux at a given energy  & $\mathrm{TeV~cm^{2}~s^{-1}}$ \\

Reviewer: ```suggestion
         e2dnde & Differential flux at a given energy  & $\mathrm{TeV~cm^{-2}~s^{-1}}$ \\
```

>          flux & Integrated flux in agiven energy range & $\mathrm{cm^{2}~s^{-1}}$ \\

Reviewer: ```suggestion
         flux & Integrated flux in a given energy range & $\mathrm{cm^{-2}~s^{-1}}$ \\
```

>          eflux & Integrated energy flux in agiven energy range & $\mathrm{erg~cm^{2}~s^{-1}}$\\

Reviewer: ```suggestion
         eflux & Integrated energy flux in a given energy range & $\mathrm{erg~cm^{-2}~s^{-1}}$\\
```

         \hline
        \end{tabular}
    \end{center}
    \label{tab:sed_types}
    \caption{Definition of the different SED types supported in \gammapy.}
\end{table*}

> Both result objects support the possibility to serialise

Reviewer: ```suggestion
Both results and objects support the possibility to serialise
```

the data into multiple formats. This includes the

> \gadf SED (reference?) format, FITS based ND sky-maps

Reviewer: https://gamma-astro-data-formats.readthedocs.io/en/v0.2/spectra/flux_points/index.html

> \gadf SED (reference?) format, FITS based ND sky-maps

Reviewer: ND=?
sky-maps -> sky maps

and formats compatible with Astropy's \code{Table} and
\code{BinnedTimeSeries} data structures. This allows
users to directly further analyse the results, e.g.
standard algorithms for time analysis such as
computing Lomb-Scargle periodograms or Bayesian
blocks for time series. So far \gammapy does not

> support "unfolding" of \gammaray spectra.

Reviewer: ```suggestion
support 'unfolding' of \gammaray spectra.
```

> Methods for this will be implemented in future version of \gammapy.

Reviewer: ```suggestion
Methods for this will be implemented in a future version of \gammapy.
```

Code example ~\ref{fig*:minted:gp_estimators} shows how to use
the \code{TSMapEstimator} objects with a given input \code{MapDataset}.
In addition to the model it allows to specify the energy
bins of the resulting flux and TS maps.

\begin{figure}
	\import{code-examples/generated/}{gp_estimators}
	\caption{Using the \code{TSMapEstimator} object from \code{gammapy.estimators} to compute a
        a flux, flux upper limits and TS map. The additional parameters \code{n\_sigma}
        and \code{n\_sigma\_ul} define the confidence levels (in multiples of the normal distribution width)
        of the flux error and and flux upper limit maps respectively.
    }
    \label{fig*:minted:gp_estimators}
\end{figure}

\subsection{gammapy.analysis}
\label{ssec:gammapy-analysis}

> The \code{gammapy.analysis} sub-module provides a  high-level interface for the most

Reviewer: not all the most common, is not it?
If not, I propose "for common use cases" ... being more quantitative

> The \code{gammapy.analysis} sub-module provides a  high-level interface for the most

Reviewer: ```suggestion
The \code{gammapy.analysis} sub-module provides a high-level interface (HLI) for the most
```

> common use cases identified in \gammaray analyses. The classes and methods
> included can be used in Python scripts, notebooks or as commands within IPython

Reviewer: ```suggestion
common use cases identified in \gammaray analyses. The included classes and methods
 can be used in Python scripts, notebooks or as commands within IPython
```

> sessions. The high-level user interface can also be used to automatise

Reviewer: ```suggestion
sessions. The HLI can also be used to automatise
```

> workflows driven by parameters declared in a configuration file in YAML format.

Reviewer: ```suggestion
workflows driven by parameters declared into a configuration file in YAML format.
```

> This way a full analysis can be executed via a single command line taking the

Reviewer: ```suggestion
This way, a full analysis can be executed via a single command line taking the
```

configuration file as input.

The \code{Analysis} class has the responsibility of orchestrating of the workflow
defined in the configuration \code{AnalysisConfig} objects and triggering the execution of

> the \code{AnalysisStep} classes that defines the common use cases identified. These

Reviewer: ```suggestion
the \code{AnalysisStep} classes that define the identified common use cases. These
```

> steps include the following: observations selection from the DataStore,  data

Reviewer: ```suggestion
steps include the following: observations selection with the \code{DataStore},  data
```

> reduction, excess map computation, model fitting, flux-points estimation, and

Reviewer: ```suggestion
reduction, excess map computation, model fitting, flux points estimation, and
```

> light-curves production. The structure of the \code{Analysis} class allows to define

Reviewer: ```suggestion
light curves production. The structure of the \code{Analysis} class allows to define
```

> custom AnalysisStep, so ones could design and configure arbitrarily complex

Reviewer: ```suggestion
custom AnalysisStep, so one could design and configure arbitrarily complex
```

> wokflows. \todo{the latter is not true for v1.0..., include example YAML file?}

Reviewer: ```suggestion
workflows. \todo{the latter is not true for v1.0..., include example YAML file?}
```

\subsection{gammapy.stats}
\label{ssec:gammapy-stats}
The \code{gammapy.stats} subpackage contains the fit statistics and associated
statistical estimators that are commonly used in \gammaray astronomy. In
general, \gammaray observations count Poisson-distributed events at various sky
positions, and contain both signal and background events. Estimation of the
number of signal events is done through likelihood maximization. In \gammapy,
the fit statistics are Poisson log-likelihood functions normalized like
chi-squares, i.e., they follow the expression $2 \time\cdot \log{\mathcal{L}}$,
where $\mathcal{L}$ is

> the likelihood function used. The statistic function used when the expected
> number of background events is known is the \emph{Cash} statistic

Reviewer: ```suggestion
the likelihood function used. When the expected number of background events is known, the used statistic function 
 is the \emph{Cash} statistic
```

~\citep{Cash}. It is used by datasets using background templates such as the

> \code{MapDataset}. When the number of background events is unknown and an off

Reviewer: ```suggestion
\code{MapDataset}. When the number of background events is unknown and an OFF
```

measurement where only background events are expected is used, the statistic

> function is WStat. It is a profile log-likelihood statistic where background

Reviewer: ```suggestion
function is \emph{WStat}. It is a profile log-likelihood statistic where background
```

> function is WStat. It is a profile log-likelihood statistic where background

Reviewer: no reference for WStat (as for the Cash)?

Author: As far as I know there is no official reference for `wstat`, we have always linked to some spec page for this....maybe @registerrier knows better?

counts are marginalized parameters. It is used by datasets containing off
counts measurements such as the \code{SpectrumDatasetOnOff}, used for classical
spectral analysis.

To perform simple statistical estimations on counts measurements,

> CountsStatistic classes encapsulate the aforementioned statistic functions to

Reviewer: ```suggestion
\code{CountsStatistic} classes encapsulate the aforementioned statistic functions to
```

measure excess counts and estimate the associated statistical significance,
errors and upper limits. They perform maximum likelihood ratio tests to
estimate significance (the square root of the statistic difference) and compute
likelihood profiles to measure errors and upper limits. The code example
\ref{codeexample:stats} shows how to compute the Li \& Ma
significance~\citep{LiMa} of a set of measurements.

\begin{figure}
	\import{code-examples/generated/}{gp_stats}
	\caption{
        Using \code{gammapy.stats} to compute statistical quantities
        such as excess, signficance and assymetric errors
        from counts based data. The data is passed on initialisation
        of the \code{WStatCountsStatistic} class. The quantities

>         are the computed on access of the corresponding class

Reviewer: ```suggestion
        are the computed ON excess of the corresponding class
```

        attributes such as \code{stat.n\_sig} and \code{stat.sqrt\_ts}.
    }
	\label{codeexample:stats} \end{figure}

\subsection{gammapy.visualization}
\label{ssec:gammapy-visualization}
The \code{gammapy.visualization} sub-package contains helper functions
for plotting and visualizing analysis results and \gammapy~data structures.
This includes for example the visualization of reflected background regions across
multiple observations or plotting large parameter correlation matrices of
\gammapy models. It also includes a helper class to split
wide field Galactic survey images across multiple panels to fit a standard
paper size.

> The sub-package also provides matplotlib implementations of specific

Author: I'll commit this change, but I agree we should introduce a common formatting for all software packages we refer to...

colormaps for false color image representation. Those colormaps have

> been used historically larger collaborations in the very high energy

Author: Addressed

> domain (such as \milagro or \hess) as "trademark" colormaps.

Author: Done

While we explicitly discourage the use of those colormaps for publication
of new results, because they do not follow modern visualization
standards, such as linear brightness gradients and accessibility
for visually impaired people, we still consider the colormaps
useful for reproducibility of past results.

\subsection{gammapy.astro}
\label{ssec:gammapy-astro}
The \code{gammapy.astro} sub-package contains utility functions for studying physical

> scenarios in high energy astrophysics. The \code{gammapy.astro.darkmatter} module

Reviewer: ```suggestion
scenarios in high-energy astrophysics. The \code{gammapy.astro.darkmatter} module
```

> computes the so  called J-factors and the associated \gammaray spectra expected

Reviewer 2: ```suggestion
computes the so called J-factors and the associated \gammaray spectra expected
```

Author: Fixed

from annihilation of dark matter in different channels according to the recipe
described in \cite{2011JCAP...03..051C}.

In the \code{gammapy.astro.source} sub-module, dedicated classes exist for modeling
galactic \gammaray sources according to simplified physical models, eg: SNR evolution

> models \citep{1950RSPSA.201..159T, 1999ApJS..120..299T}, evolution of PWN during the

Reviewer 2: SNR and PWN acronym not defined

> models \citep{1950RSPSA.201..159T, 1999ApJS..120..299T}, evolution of PWN during the

Author: Done

free expansion phase \citep{2006ARA&A..44...17G} or computation
of physical parameters in a pulsar assumed to be a simple dipole.

In the \code{gammapy.astro.population} sub-module there are dedicated tools
for simulating synthetic populations based on physical models derived from
observational or theoretical considerations for different classes of Galactic

> VHE \gammaray emitters: PWNe, SNRs \cite{1998ApJ...504..761C},

Reviewer 2: VHE acronym not defined

pulsars \cite{2006ApJ...643..332F, 2006MNRAS.372..777L, 2004A&A...422..545Y}
and \gammaray binaries.

While the present list of use cases is rather preliminary, this can be enriched
with time with by users and/or developers according to future needs.

\subsection{gammapy.catalog}
\label{ssec:gammapy-catalog}

> Comprehensive source catalogs are increasingly being provided by many high
> energy astrophysics experiments. The \code{gammapy.catalog} sub-packages

Reviewer: ```suggestion
Comprehensive source catalogs are increasingly being provided by many high-energy
 astrophysics experiments. The \code{gammapy.catalog} sub-packages
```

provides a convenient access to the most important \gammaray catalogs.
Catalogs are represented by the \code{SourceCatalog} object, which
contains the actual catalog as an Astropy \code{Table} object.
Objects in the catalog can be accesed by row index, name of the
object or any association or alias name listed in the catalog.

Sources are represented in \gammapy by the \code{SourceCatalogObject}
class, which has the responsibility to translate the information
contained in the catalog to other \gammapy objects. This includes

> the spatial and spectral model of the source, flux points and

Reviewer: ```suggestion
the spatial and spectral models of the source, flux points and
```

> light curves (if available) for individual objects. This

Reviewer: ```suggestion
light curves (if available) for each individual object. This
```

module works independently from the rest of the package, and the required
catalogs are supplied in \code{GAMMAPY\_DATA} repository.
The overview of currenly supported catalogs, the corresponding
\gammapy classes and references are shown in Table~\ref{tab:catalogs}.
Newly released relevant catalogs will be added in future.

\begin{table*}[ht!]
    \begin{center}
        \begin{tabular}{llll}
         \hline
         Class Name & Shortcut & Description & Reference\\
         \hline
         \code{SourceCatalog3FGL} & \code{"3fgl"} & 3\textsuperscript{rd} catalog of \fermi sources & \cite{3FGL} \\
         \code{SourceCatalog4FGL} & \code{"4fgl"} & 4\textsuperscript{th} catalog of \fermi  sources & \cite{4FGL} \\

>          \code{SourceCatalog2FHL} & \code{"2fhl"} & 2\textsuperscript{nd} catalog high energy \fermi  sources & \cite{2FHL} \\

Reviewer: ```suggestion
         \code{SourceCatalog2FHL} & \code{"2fhl"} & 2\textsuperscript{nd} catalog high-energy \fermi  sources & \cite{2FHL} \\
```

>          \code{SourceCatalog3FHL} & \code{"3fhl"} & 3\textsuperscript{rd} catalog high energy \fermi  sources & \cite{3FHL} \\

Reviewer: ```suggestion
         \code{SourceCatalog3FHL} & \code{"3fhl"} & 3\textsuperscript{rd} catalog high-energy \fermi  sources & \cite{3FHL} \\
```

         \code{SourceCatalog2HWC} & \code{"2hwc"} & 2\textsuperscript{nd} catalog of \hawc sources & \cite{2HWC} \\
         \code{SourceCatalog3HWC} & \code{"3hwc"} & 3\textsuperscript{rd} catalog of \hawc sources & \cite{3HWC} \\
         \code{SourceCatalogHGPS} & \code{"hgps"} & \hess Galactic Plane Survey catalog & \cite{HGPS} \\

>          \code{SourceCatalogGammaCat} & \code{"gammacat"} & Open source data collection & \citep{gamma-cat} \\

Reviewer: gammacat is not maintained and a bit out-dated. Should we really mention it??

>          \code{SourceCatalogGammaCat} & \code{"gammacat"} & Open source data collection & \citep{gamma-cat} \\

Reviewer 2: ```suggestion
         \code{SourceCatalogGammaCat} & \code{"gammacat"} & Open source data collection & \cite{gamma-cat} \\
```

         \hline
         \end{tabular}
    \end{center}
    \label{tab:catalogs}
    \caption{Overview of supported catalogs in \code{gammapy.catalog}.}
\end{table*}

\begin{figure}
	\import{code-examples/generated/}{gp_catalogs}
	\caption{Using \code{gammapy.catalogs} to access the underlying model, flux points and
		light-curve from the \fermi 4FGL catalog for the blazar PKS 2155-304}
	\label{codeexample:data}
\end{figure}

> %\subsection{gammapy.utils}
> \label{ssec:gammapy-utils}
> Utility functions...

Reviewer 2: ```suggestion
%\subsection{gammapy.utils}
%\label{ssec:gammapy-utils}
%Utility functions...
```

\section{Applications}
\label{sec:applications}

\gammapy can be used for a variety of science cases by different IACT

> experiments. (Refer to publications) In this section, we show a non-exhaustive

Reviewer: which publications????

list of some typical analysis that can be performed. In general from the
\gammapy side there is not limitation on which kind of analysis can be done
with which instrument, however in practice it is limited by the availability of
public data.

\subsection{1D Analysis}
\label{ssec:1d-analysis}
One of the most common analysis cases in \gammaray astronomy is measuring the
spectrum of a source in a given region defined on the sky, in conventional
astronomy also called aperture photometry. The spectrum is typically measured
in two steps: first a parametric spectral model is fitted to the data and
secondly flux points are computed in a pre-defined set of energy bins. The
result of such an analysis performed on three simulated CTA observations is
shown in Fig.~\ref{fig:cta_galactic_center}. In this case the spectrum was
measured in a circular aperture centered on the Galactic Center, in

> \gammaray~astronomy often called "on region". For such analysis the users first

Reviewer: ```suggestion
\gammaray~astronomy often called 'ON region'. For such analysis the user first
```

chooses a region of interest and energy binning, both defined by a

> \code{RegionGeom}. In a second step the events and instrument response are binned

Reviewer: ```suggestion
\code{RegionGeom}. In a second step the events and the IRFs are binned
```

into maps of this geometry, by the \code{SpectrumDatasetMaker}. All the data and

> reduced instrument response are bundled into a \code{SpectrumDataset}. To estimate

Reviewer: ```suggestion
reduced IRFs are bundled into a \code{SpectrumDataset}. To estimate
```

> the expected background in the on region a "reflected regions" background

Reviewer: ```suggestion
the expected background in the ON region a 'reflected regions' background
```

method was used~\cite{Berge07}, represented in \gammapy by the
\code{ReflectedRegionsBackgroundMaker} class. The resulting reflected regions are
illustrated for all three observations on top of the map of counts. After

> reduction of the data it was modelled using a forward-folding method and a

Reviewer: what is 'it'? the ON region counts stored into the SpectrumDataset?

> power-law spectral shape, using the \code{PowerLawSpectralModel} and \code{Fit} class.

Reviewer: ```suggestion
power law spectral shape, using the \code{PowerLawSpectralModel} and \code{Fit} class.
```

Based on this best fit model the final flux points and corresponding
log-likelihood profiles are computed using the \code{FluxPointsEstimator}.

\begin{figure*}[t]
	\centering
	\includegraphics[width=1.\textwidth]{figures/cta_galactic_center.pdf}
	\caption{

> 		Example spectral analysis of the Galactic Center for three simulated CTA

Reviewer: ```suggestion
		Example of 1D spectral analysis of the Galactic Center for three simulated CTA
```

		observations. The left image shows the maps of counts with the measurement
		region and background regions overlaid in different colors. The right image
		shows the resulting spectral points and their corresponding log-likelihood
		profiles.} \label{fig:cta_galactic_center} \end{figure*}

> \subsection{3D Analysis}

Reviewer: No reference to the Fermi analysis?

\label{ssec:3d-analysis}
The 1D analysis approach is a powerful tool to measure the spectrum of an
isolated source. However, more complicated situations require a more careful
treatment. In a field of view containing several overlapping sources, the 1D
approach cannot disentangle the contribution of each source to the total flux in
the selected region. Sources with extended or complex morphology can result in
the measured flux being underestimated, and heavily dependent on the choice of
extraction region. Additionally, the 1D approach neglects the energy-dependence

> of the PSF.

Reviewer: This is not true! the PSF is taking into account for the leakage. I propose to remove this sentence.

For such situations a more complex approach is needed, the so-called 3D
analysis. The three relevant dimensions are the two spatial angular coordinates
and an energy axis. In this framework, a combined spatial and spectral model
(that is, a \code{SkyModel}, see Section~\ref{ssec:gammapy-modeling}) is fitted to the

> sky-maps that were previously derived from the data and bundled into a

Reviewer: ```suggestion
sky maps that were previously derived from the data reduction step and bundled into a
```

\code{MapDataset} (see Sections~\ref{ssec:gammapy-makers} and~\ref{ssec:gammapy-datasets}).

A thorough description of the 3D analysis approach and multiple examples that
use \gammapy can be found in~\cite{Mohrmann2019}. Here we present a short
example to highlight some of its advantages.

Starting from the \irfs corresponding to the same three simulated \cta
observations used in Section~\ref{ssec:1d-analysis}, we can create a \code{MapDataset}
via the \code{MapDatasetMaker}. However, we will not use the simulated event lists

> provided by \cta but instead, use the method MapDataset.fake() to simulate

Reviewer: ```suggestion
provided by \cta but instead, use the method \code{MapDataset.fake()} to simulate
```

> measured counts from the combination of several SkyModel instances. In this

Reviewer: ```suggestion
measured counts from the combination of several \code{SkyModel} instances. In this
```

way, a DL4 dataset can directly be simulated. In particular we simulate:

\begin{enumerate}

>     \item A point source located at (l=0\textdegree, b=0\textdegree) with a power-law

Reviewer: ```suggestion
    \item A point source located at (l=0\textdegree, b=0\textdegree) with a power law
```

	      spectral shape.
    \item An extended source with Gaussian morphology located at (l=0.4\textdegree,
	      b=0.15\textdegree) with $\sigma$=0.2\textdegree and a log-parabola spectral
	      shape.
    \item A large shell-like structure centered around (l=0.06\textdegree,
	      b=0.6\textdegree) with a radius and width of 0.6\textdegree and 0.3\textdegree

> 	      respectively and a power-law spectral shape.

Reviewer: ```suggestion
	      respectively and a power law spectral shape.
```

\end{enumerate}

The position and sizes of the sources
have been selected so that their contributions overlap. This can be clearly
seen in the significance map shown in the left panel of
Figure~\ref{fig:cube_analysis}. This map was produced with the
\code{ExcessMapEstimator} (see Section~\ref{ssec:gammapy-estimators}) with a
correlation radius of 0.1\textdegree.

We can now fit the same model shapes to the simulated data and retrieve the
best-fit parameters. To check the model agreement, we compute the residual
significance map after removing the contribution from each model. This is done
again via the \code{ExcessMapEstimator}. As can be seen in the middle panel of
Figure~\ref{fig:cube_analysis}, there are no regions above or below 5$\sigma$,
meaning that the models describe the data sufficiently well.

As the example above shows, the 3D analysis allows to characterize the
morphology of the emission and fit it together with the spectral properties of
the source.  Among the advantages that this provides is the ability to
disentangle the contribution from overlapping sources to the same spatial

> region. To highlight this, we define a circular RegionGeom of radius

Reviewer: ```suggestion
region. To highlight this, we define a circular \code{RegionGeom} of radius
```

Author: Done

0.5\textdegree~ centered around the position of the point source, which is drawn
in the left panel of Figure~\ref{fig:cube_analysis}. We can now compare the
measured excess counts integrated in that region to the expected relative
contribution from each of the three source models. The result can be seen in the left
panel of Figure~\ref{fig:cube_analysis}.

\begin{figure*}[t]
	\centering
	\includegraphics[width=1.\textwidth]{figures/cube_analysis.pdf}

> 	\caption{Example 3D analysis for simulated sources using the \cta \irfs. The

Reviewer: ```suggestion
	\caption{Example of a 3D analysis for simulated sources using the \cta \irfs. The
```

		left image shows a significance map where the three simulated sources can be
		seen. The middle figure shows another significance map, but this time after
		subtracting the best-fit model for each of the sources, which are displayed in
		black. The right figure shows the contribution of each source model to the
		circular region of radius 0.5\textdegree drawn in the left image, together with
		the excess counts inside that region. } \label{fig:cube_analysis}
\end{figure*}

Note that all the models fitted also have a spectral component, from which flux
points can be derived in a similar way as described in~\ref{ssec:1d-analysis}.
%\end{figure*}%	\caption{Fermi-LAT TS map in two energy bands} \label{fig:fermi_ts_map}%	\includegraphics[width=1.\textwidth]{figures/fermi_ts_map.pdf}%Ref:~\citep{Stewart2009} \begin{figure*}[t] \centering%\todo{What to do with } Figure~\ref{fig:fermi_ts_map} ?%

\subsection{Temporal Analysis}
\label{ssec:temporal-analysis}
A common use case in most astrophysical scenarios is to study the temporal

> variability of the source. The most basic way to do this is to construct a

Author: Done

> lightcurve, i.e., the flux of the source in each given time bin. In \gammapy, this

Author: Done

> is done by using the \code{LightCurveEstimator} which fits the normalisation of the

Author: Done

> source in each energy band per observation, keeping other parameters constant.

Author: Clarified

> source in each energy band per observation, keeping other parameters constant.

Author: Done

IRF is performed to predict the counts in each estimated energy bin. A log
parabola is fitted to the almost fived decades in energy $10\,{\rm GeV} --
	300\,{\rm TeV}$. %% is the LP formula specified elsewhere?

The result of the joint fit is displayed in
Fig.~\ref{fig:multi_instrument_analysis}. We remark that the objective of this
exercise is illustrative, we display the flexibility of \gammapy in
simultaneously fitting multi-instrument data even at different levels of
reduction, we do not aim to provide a new measurement of the Crab Nebula
spectrum.

%% eventual part for physical modelling with naima

\subsection{Broadband SED Modeling}
\label{ssec:broadband-sed-modeling}
By combining \gammapy with astrophysical modelling codes users can also fit
astrophysical spectral models to \gammaray data. In \gammaray
astronomy what typically observes two radiation production
mechanisms., the so called hadronic and leptonic scenarios.
There are mutiple Python packages that are able to model
the \gammaray emission, given a physical scenario. Among those
packages are  \cite{agnpy}, \cite{naima}, \cite{jetset} or \cite{gamera}.
Tyically those emission models predict broadband emission from

> radio waves up to the very high energy \gammaray range.

Reviewer: ```suggestion
radio waves up to the very high-energy \gammaray range.
```

four main steps : data-reduction; object detection; model fitting and model
selection; associations and classification. The \iacts data-reduction step is
done in the same way than described in the previous sections but scale up to
few thousand of observations. The object detection step consists a minima in
finding local maxima in the significance, or TS maps,  given by the
\code{ExcessMapEstimator}, or \code{TSMapEstimator}, respectively.  Further refinements can
include for example  filtering and detection on these maps with techniques from
scikit-image package~\citep{scikit-image}, and outlier detection from
scikit-learn package~\citep{scikit-learn}. Tests on simulations shown that it
reduces the spurious detections at this stage and then speed up the next step
as less objects will have to be fitted simultaneously. During the modelling
step each object is alternatively fitted with different models in order to
determine their optimal parameters, and the best-candidate model. The
subpackage \code{gammapy.modeling.models} offers a large variety of choice, and the
possibility to add custom models.  Several spatial models (point-source, disk,

> gaussian...), and spectral models (power-law, log-parabola...) may be tested

Reviewer: ```suggestion
Gaussian...), and spectral models (power law, log-parabola...) may be tested
```

for each object, so the complexity of the problem increases rapidly in regions
crowded with multiple extended sources. Finally an object is discarded if its
best-fit model is not significantly preferred over the null hypothesis (no

> source) comparing the difference in log-Likelihood between these two

Reviewer: ```suggestion
source) comparing the difference in log likelihood between these two
```

hypotheses. For the association and classification step, that is tightly
connected to the population studies, we can a minima compare the fitted models

> to the set of gammapy-ray catalogs available in gammapy.catalog. However

Reviewer: ```suggestion
to the set of \gammaray catalogs available in \code{gammapy.catalog}. However
```

Author: Done
